### PR TITLE
refactor: Getting rid of (un)assigningCb

### PR DESCRIPTION
**Pull request by dorjesinpo**
One more step converging CSL/non-CSL

### CODE REVIEW

File: src/groups/mqb/mqbc/mqbc_clusterutil.cpp [hunk 1]
```diff
@@ -897,6 +896,20 @@ ClusterUtil::assignQueue(ClusterState*         clusterState,
             allocator);
         domIt = clusterState->domainStates().find(uri.qualifiedDomain());
     }
+    else {
+        // Set the queue as assigning (no longer pending unassignment)
+
+        UriToQueueInfoMapCIter qcit = domIt->second->queuesInfo().find(uri);
+        if (qcit != domIt->second->queuesInfo().cend()) {
```
Comment (kaikulimu) [inline src/groups/mqb/mqbc/mqbc_clusterutil.cpp:903]: This approach doesn't actually work. Assignment is a different story than unassignment. During first pending assignment, the queue does not exist in `ClusterState` yet, so `qcit` won't find anything, and thus won't set anything to `k_ASSIGNING`. 

*Lemma*: If the queue exists in `ClusterState`, then it is never in `k_ASSIGNING` state.

Therefore, we need a separate collection to keep track of `k_ASSIGNING` queues, and once assignment at `ClusterState` completes, we need to remove the corresponding entry.

**Comment by kaikulimu** [comment]
Some notes for myself

`ClusterQueueHelper::onQueueAssigned`:
    - Legacy mode: Same observer callback, but skipped due to immediate return
    - FSM mode: Observer callback of `ClusterState::assignQueue`

`ClusterQueueHelper::onQueueAssigning`:
    - Legacy mode: Called during `ClusterStateManager::assignQueue`, `ClusterStateManager::registerQueueInfo`, `ClusterStateManager::processQueueAssignmentRequest`, `ClusterStateManager::processQueueAssignmentAdvisory`
    - FSM mode: Called during `ClusterStateManager::assignQueue`, `ClusterStateManager::registerQueueInfo`, `ClusterStateManager::processQueueAssignmentRequest`
    > Both modes share the same first three codepaths via `ClusterUtil`
    > `ClusterStateManager::processQueueAssignmentRequest` simply fowards to `ClusterUtil::assignQueue`
    > `mqbblp::ClusterStateManager::processQueueAssignmentAdvisory` is unique to legacy mode

`ClusterQueueHelper::onQueueUnassigned`:
    - Legacy mode: Same observer callback, but skipped due to immediate return
    - FSM mode: Observer callback of `ClusterState::unassignQueue`
    
`ClusterQueueHelper::onQueueUnassigning`:
    - Legacy mode: Called during [`ClusterStateManager::processQueueUnAssignmentAdvisory`](https://github.com/bloomberg/blazingmq/blob/main/src/groups/mqb/mqbblp/mqbblp_clusterstatemanager.cpp#L1954)
    - FSM mode: no-op
    > In legacy mode, on the leader, `removeQueueRaw` is called during `ClusterQueueHelper::gcExpiredQueue`

**Comment by kaikulimu** [comment]
Reviewing, should be done by Dev 5

**Comment by kaikulimu** [comment]
Progress update:
Queue unassignment flow has been reviewed.
Now moving on to queue assignment flow.

I have refrained from posting my review, in case my perception on queue unassignment aspects changes after I review queue assignment flow.

File: src/groups/mqb/mqbblp/mqbblp_clusterqueuehelper.cpp [hunk 2]
```diff
@@ -4611,43 +4452,6 @@ void ClusterQueueHelper::openQueue(
 
     QueueContextMapIter queueContextIt = d_queues.find(uriKey);
 
-    // NOTE: See TBD in 'onGetDomainDispatched': if the request comes from a
```
Comment (kaikulimu) [inline src/groups/mqb/mqbblp/mqbblp_clusterqueuehelper.cpp:4614]: To be honest, I am quite unfamiliar with logic surrounding 'clientIdentity'. Are you sure it is safe to remove this code block?

File: src/groups/mqb/mqbblp/mqbblp_clusterqueuehelper.cpp [hunk 3]
```diff
@@ -4328,7 +4177,9 @@ void ClusterQueueHelper::onQueueUnassigned(
                     -1);
             }
             d_queuesById.erase(qinfo.d_id);
-            qinfo.reset();
+            qinfo.resetAndKeepPending();
+            // CQH will recreate 'queueContextSp->d_liveQInfo.d_queue_sp' upon
+            // 'onOpenQueueResponse'
```
[the remark this answers is not in the document]
Comment (kaikulimu) [reply src/groups/mqb/mqbblp/mqbblp_clusterqueuehelper.cpp:4182]: A similar situation occurs during [`ClusterQueueHelper::onQueueAssigned`](https://github.com/bloomberg/blazingmq/blob/ea72c8e4362c0e467531139202cc2b8a697904f8/src/groups/mqb/mqbblp/mqbblp_clusterqueuehelper.cpp#L4031-L4061) in this PR, where you decided to put `if (d_cluster_p->isCSLModeEnabled()) {}` around the block.

This is a satisfactory solution too, if you don't want to refactor too much more code.

**Review by kaikulimu — CHANGES_REQUESTED** [review]
Review done. Thanks for the hard work! This part of the code is not easy to navigate.

File: src/groups/mqb/mqbc/mqbc_clusterutil.cpp [hunk 4]
```diff
@@ -1174,27 +1158,6 @@ void ClusterUtil::registerQueueInfo(ClusterState*           clusterState,
...
-                << BMQTSK_ALARMLOG_END;
-            return;  // RETURN
-        }
-
-        clusterState->domainStates()
-            .at(uri.qualifiedDomain())
-            ->adjustQueueCount(1);
-    }
-
-    queueAssigningCb(uri, false);  // processingPendingRequests
```
[the remark this answers is not in the document]
Comment (dorjesinpo) [reply src/groups/mqb/mqbc/mqbc_clusterutil.cpp:1197]: That is the main purpose of this PR.  We want to minimize and straighten the assignment logic.

File: src/groups/mqb/mqbblp/mqbblp_clusterqueuehelper.cpp [hunk 5]
```diff
@@ -4159,22 +4009,19 @@ void ClusterQueueHelper::onQueueAssigned(
...
         queueContext.reset(new (*d_allocator_p)
                                QueueContext(info.uri(), d_allocator_p),
                            d_allocator_p);
 
         d_queues[info.uri()] = queueContext;
     }
+    mqbc::ClusterState::DomainState& domainState =
+        *d_clusterState_p->domainStates().at(info.uri().qualifiedDomain());
+
+    domainState.adjustQueueCount(1);
```
[the remark this answers is not in the document]
Comment (dorjesinpo) [reply src/groups/mqb/mqbblp/mqbblp_clusterqueuehelper.cpp:4023]: Bad idea, actually,  `ClusterState::assignQueue` can be called multiple times leading to a miscounting
See `test_max_queue_restore[multi_node_fsm]`

**Review by kaikulimu — APPROVED** [review]
lgtm

**Pull request merged**